File: services/profile_service.go
package services

import (
	"context"
	"errors"
	"time"

	"doctor-service/dto"
	"doctor-service/repositories"

	"github.com/google/uuid"
)

type ProfileService interface {
	UpdateProfile(ctx context.Context, doctorID uuid.UUID, req dto.UpdateProfileRequest) (*dto.DoctorProfileDTO, error)
	GetProfile(ctx context.Context, doctorID uuid.UUID) (*dto.DoctorProfileDTO, error)
}

type DefaultProfileService struct {
	doctorRepo repositories.DoctorRepository
}

func NewProfileService(doctorRepo repositories.DoctorRepository) ProfileService {
	return &DefaultProfileService{doctorRepo: doctorRepo}
}

func (s *DefaultProfileService) UpdateProfile(ctx context.Context, doctorID uuid.UUID, req dto.UpdateProfileRequest) (*dto.DoctorProfileDTO, error) {
	doc, err := s.doctorRepo.FindByPublicID(ctx, doctorID)
	if err != nil || doc == nil {
		return nil, errors.New("doctor profile not found")
	}

	if req.Gender != "" {
		doc.Gender = req.Gender
	}
	if req.ProfilePhoto != "" {
		doc.ProfilePhoto = req.ProfilePhoto
	}
	if req.Languages != "" {
		doc.Languages = req.Languages
	}
	if req.Biography != "" {
		doc.Biography = req.Biography
	}
	if req.DOB != "" {
		parsed, err := time.Parse("2006-01-02", req.DOB)
		if err == nil {
			doc.DOB = &parsed
		}
	}

	if err := s.doctorRepo.Update(ctx, doc); err != nil {
		return nil, err
	}

	return &dto.DoctorProfileDTO{
		PublicID:       doc.PublicID,
		FirstName:      doc.FirstName,
		LastName:       doc.LastName,
		Email:          doc.Email,
		Mobile:         doc.Mobile,
		Status:         string(doc.Status),
		MobileVerified: doc.MobileVerified,
		EmailVerified:  doc.EmailVerified,
	}, nil
}

func (s *DefaultProfileService) GetProfile(ctx context.Context, doctorID uuid.UUID) (*dto.DoctorProfileDTO, error) {
	doc, err := s.doctorRepo.FindByPublicID(ctx, doctorID)
	if err != nil || doc == nil {
		return nil, errors.New("doctor profile not found")
	}

	return &dto.DoctorProfileDTO{
		PublicID:       doc.PublicID,
		FirstName:      doc.FirstName,
		LastName:       doc.LastName,
		Email:          doc.Email,
		Mobile:         doc.Mobile,
		Status:         string(doc.Status),
		MobileVerified: doc.MobileVerified,
		EmailVerified:  doc.EmailVerified,
	}, nil
}

File: python_backend/services/profile_service.py
import uuid
from datetime import datetime

class ProfileService:
    def UpdateProfile(self, ctx, doctor_id, req): pass
    def GetProfile(self, ctx, doctor_id): pass

class DefaultProfileService(ProfileService):
    def __init__(self, doctor_repo):
        self.doctorRepo = doctor_repo

    def UpdateProfile(self, ctx, doctor_id, req):
        doc = self.doctorRepo.FindByPublicID(ctx, doctor_id)
        if not doc:
            raise Exception("doctor profile not found")

        if getattr(req, 'Gender', None):
            doc.Gender = req.Gender
        if getattr(req, 'ProfilePhoto', None):
            doc.ProfilePhoto = req.ProfilePhoto
        if getattr(req, 'Languages', None):
            doc.Languages = req.Languages
        if getattr(req, 'Biography', None):
            doc.Biography = req.Biography
        if getattr(req, 'DOB', None):
            try:
                doc.DOB = datetime.strptime(req.DOB, "%Y-%m-%d")
            except ValueError:
                pass

        try:
            self.doctorRepo.Update(ctx, doc)
        except Exception as e:
            raise e

        class DoctorProfileDTO: pass
        resp = DoctorProfileDTO()
        resp.PublicID = doc.PublicID
        resp.FirstName = doc.FirstName
        resp.LastName = doc.LastName
        resp.Email = doc.Email
        resp.Mobile = doc.Mobile
        resp.Status = str(doc.Status)
        resp.MobileVerified = doc.MobileVerified
        resp.EmailVerified = doc.EmailVerified
        return resp

    def GetProfile(self, ctx, doctor_id):
        doc = self.doctorRepo.FindByPublicID(ctx, doctor_id)
        if not doc:
            raise Exception("doctor profile not found")

        class DoctorProfileDTO: pass
        resp = DoctorProfileDTO()
        resp.PublicID = doc.PublicID
        resp.FirstName = doc.FirstName
        resp.LastName = doc.LastName
        resp.Email = doc.Email
        resp.Mobile = doc.Mobile
        resp.Status = str(doc.Status)
        resp.MobileVerified = doc.MobileVerified
        resp.EmailVerified = doc.EmailVerified
        return resp
